### PR TITLE
OP_CHECKSEQUENCEVERIFY

### PR DESCRIPTION
**Pull request by natzei**
Implementation of the opcode OP_CHECKSEQUENCEVERIFY.

See [BIP 68](https://github.com/bitcoin/bips/blob/master/bip-0068.mediawiki) and [BIP 112](https://github.com/bitcoin/bips/blob/master/bip-0112.mediawiki) for details.

Class ScriptTest has been enhanced to support transaction setup via json (tx version and input sequence).

### CODE REVIEW

File: core/src/test/java/org/bitcoinj/script/ScriptTest.java [hunk 1]
```diff
@@ -263,7 +263,7 @@ private Script parseScriptString(String string) throws IOException {
                 // opcode, e.g. OP_ADD or OP_1:
                 out.write(ScriptOpCodes.getOpCode(w.substring(3)));
             } else {
-                throw new RuntimeException("Invalid Data");
+                throw new RuntimeException("Invalid Data '"+w+"'");
```
Comment (schildbach) [inline core/src/test/java/org/bitcoinj/script/ScriptTest.java]: What's that exception message?

File: core/src/test/resources/org/bitcoinj/script/script_invalid.json [hunk 2]
```diff
@@ -151,8 +151,8 @@
 ["2 2 LSHIFT", "8 EQUAL", "P2SH,STRICTENC", "disabled"],
 ["2 1 RSHIFT", "1 EQUAL", "P2SH,STRICTENC", "disabled"],
 
-["1","NOP1 NOP2 NOP3 NOP4 NOP5 NOP6 NOP7 NOP8 NOP9 NOP10 2 EQUAL", "P2SH,STRICTENC"],
```
Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: I guess these tests are copied from Bitcoin Core?

File: AUTHORS [hunk 3]
```diff
@@ -50,14 +64,18 @@ Martin Zachrison
...
 Michael Bumann
 Mike Hearn
 Mike Rosseel
 Miron Cuperman
+Miron Cuperman (devrandom)
 monk
 Mora Zyx
 mruddy
+Nicola
+Nicola Atzei
```
Comment (schildbach) [inline AUTHORS]: No need to add the missing authors now, we do that before the next release. It's fine if you add yourself if you wish so.

**Comment by schildbach** [comment]
You might want to add the "CHECKSEQUENCEVERIFY tests" from tx_valid.json and tx_invalid.json as well.

File: core/src/test/resources/org/bitcoinj/script/script_invalid.json (hunk 2 above)
Comment (natzei) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: From the Scala library (bitcoin-lib). I suppose they had copied them from Bitcoin Core.

File: AUTHORS (hunk 3 above)
Comment (natzei) [inline AUTHORS]: Sorry, I didn't know it. Should I revert the commit? Or you can edit it?

Comment (schildbach) [inline AUTHORS]: If you interactively rebase your branch on master it's easy to take out or edit the commit. But sure, I can also leave it out when merging.

**Comment by natzei** [comment]
@schildbach I need to study BIP 9 before answering your question, I didn't catch it.

File: core/src/test/resources/org/bitcoinj/script/script_invalid.json (hunk 2 above)
Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: The original can be found at https://github.com/bitcoin/bitcoin/blob/master/src/test/data/script_tests.json. The file has been merged from script_valid.json and script_invalid.json in with https://github.com/bitcoin/bitcoin/commit/dde46d3ae1de18700e089bc1861cf1e18649dc5d#diff-a79f02a42f31e934e3bdb4f40500b908, something we should follow too (maybe even before merging this PR). Anyway, I suggest gathering as much CSV-related test data from the original file as possible, because your code needs to pass these tests.

Comment (natzei) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: Sure, you're right. I'll merge the two files and look at the tests there.
I've just checked the original tests about CHECKSEQUENCEVERIFY. I wonder how they deal with tx version and tx input sequence. Any hint?

Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: Before you merge the files in this PR, it would be better to do it in a separate PR and merge it before this one. I can take a look at the merging if you prefer to stay with working on the actual CSV stuff. I currently can't answer your question, have to dive deeper into it first.

Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: Regarding the merging of the script tests please see https://github.com/bitcoinj/bitcoinj/pull/1394

Comment (natzei) [inline core/src/test/resources/org/bitcoinj/script/script_invalid.json]: After I have been spending several hours comparing bitcoinj with bitcoin-core, and I concluded that the script evaluation should be refactored a little bit to cover the same tests.

I think it's independent from this PR, isn't it? I could work on filling the gap with bitcoin-core, considering it's quite important for the work I'm doing. 

Let me know how to proceed. Your repo, your rules :)

**Comment by schildbach** [comment]
Once https://github.com/bitcoinj/bitcoinj/pull/1394 is merged, you can rebase this on master and just add the CSV-related tests from Bitcoin Core to `script_tests.json`, `tx_valid.json` and `tx_invalid.json`.

**Comment by natzei** [comment]
Rebased

**Comment by schildbach** [comment]
Great, this already looks much cleaner. Note however that the two files `script_valid.json` and `script_invalid.json` are now not existing any more. I suggest going through https://github.com/bitcoin/bitcoin/blob/master/src/test/data/script_tests.json and copy the CHECKSEQUENCEVERIFY tests from there to `script_tests.json`.

**Comment by natzei** [comment]
Done!

File: core/src/test/resources/org/bitcoinj/script/script_tests.json [hunk 4]
```diff
@@ -1119,5 +1121,1489 @@
 
 ["0x00", "'00' EQUAL", "P2SH,STRICTENC", "EVAL_FALSE", "Basic OP_0 execution"],
 
+["MINIMALDATA enforcement for PUSHDATAs"],
```
Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_tests.json]: If this block, please only copy the CSV-related tests.

(e.g. MINIMALDATA would go into a different PR that fixes #1396, but don't worry about that in the context of this PR.)

**Comment by natzei** [comment]
I'm becoming crazy.
```
["0", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME", "CSV fails if stack top bit 1 << 31 is set and the tx version < 2"],
["4294967296", "CHECKSEQUENCEVERIFY", "CHECKSEQUENCEVERIFY", "UNSATISFIED_LOCKTIME", "CSV fails if stack top bit 1 << 31 is not set, and tx version < 2"],
```
The first test passes.
The second test fails.

However, both do **not** set the bit 1 << 31 (so the comment of the first one is wrong).

I'm already checking why the second script evaluates true.

File: core/src/main/java/org/bitcoinj/script/Script.java [hunk 5]
```diff
@@ -1421,7 +1430,84 @@ private static void executeCheckLockTimeVerify(Transaction txContainingThis, int
...
+        // Note that elsewhere numeric opcodes are limited to
+        // operands in the range -2**31+1 to 2**31-1, however it is
+        // legal for opcodes to produce results exceeding that
+        // range. This limitation is implemented by CScriptNum's
+        // default 4-byte limit.
+        //
+        // Thus as a special case we tell CScriptNum to accept up
+        // to 5-byte bignums, which are good until 2**39-1, well
+        // beyond the 2**32-1 limit of the nSequence field itself.
+        final long nSequence = castToBigInteger(stack.getLast(), 5).longValue();
```
Comment (natzei) [inline core/src/main/java/org/bitcoinj/script/Script.java]: It seems that 4294967296 (1<<32) is then converted to 2147483648 (1<<31).
I don't not if it's expected.

**Comment by natzei** [comment]
Ok. I think we are ready to merge.

File: core/src/test/resources/org/bitcoinj/script/script_tests.json [hunk 6]
```diff
@@ -349,7 +349,7 @@
 ["2147483647", "0x04 0xFFFFFF7F EQUAL", "P2SH,STRICTENC", "OK"],
 ["2147483648", "0x05 0x0000008000 EQUAL", "P2SH,STRICTENC", "OK"],
 ["549755813887", "0x05 0xFFFFFFFF7F EQUAL", "P2SH,STRICTENC", "OK"],
-["549755813888", "0x06 0xFFFFFFFF7F EQUAL", "P2SH,STRICTENC", "OK"],
+["549755813888", "0x06 0x000000008000 EQUAL", "P2SH,STRICTENC", "OK"],
```
Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_tests.json]: Just out of curiosity, is this change related to CSV?

File: core/src/test/resources/org/bitcoinj/script/script_tests.json [hunk 7]
```diff
@@ -1098,7 +1099,7 @@
 "NOP NOP NOP NOP NOP NOP NOP NOP NOP NOP NOP NOP NOP 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG 0 0 'a' 'b' 'c' 'd' 'e' 'f' 'g' 'h' 'i' 'j' 'k' 'l' 'm' 'n' 'o' 'p' 'q' 'r' 's' 't' 20 CHECKMULTISIG",
 "P2SH,STRICTENC",
 "OP_COUNT",
-"Fails due to 201 sig op limit"],
+"Fails due to 201 script operation limit"],
```
Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_tests.json]: And this one?

Comment (natzei) [inline core/src/test/resources/org/bitcoinj/script/script_tests.json]: No, they aren't.

Comment (schildbach) [inline core/src/test/resources/org/bitcoinj/script/script_tests.json]: Ok no worries I'll take them out when merging.

**Comment by natzei** [comment]
In my first PULL request I used the Scala test cases and added some new.

However, to cover all the cases of CSV it's needed to set the version number and the input sequence in the spending tx. So I extended the JSON test array to specify tx informations (ignoring that these tests are "shared" between the different implementations)

Actually, Bitcoin Core (and Scala too) don't cover all the cases of CSV.

**Comment by schildbach** [comment]
I just noticed we didn't include the CSV tx_valid.json and tx_invalid.json tests in this PR. I just added those and the tests pass. Are you ok if I just include them, or would you like to take a look first?

**Comment by natzei** [comment]
I trust you :)

**Comment by schildbach** [comment]
Merged, many thanks!
